core/src/main/java/org/infinispan/interceptors/impl/CacheLoaderInterceptor.java

```diff
@@ -704,6 +707,26 @@ public boolean contains(Object o) {
          return new LocalCacheStream<>(new PersistenceEntryStreamSupplier<>(cache, iceFactory, partitioner::getSegment,
                cacheSet.stream(), persistenceManager), parallel, cache.getAdvancedCache().getComponentRegistry());
       }
+
+      @Override
+      public Publisher<CacheEntry<K, V>> localPublisher(int segment) {
+         Set<K> seenKeys = new HashSet<>(cache.getAdvancedCache().getDataContainer().sizeIncludingExpired());
+         Flowable<CacheEntry<K, V>> inMemoryPublisher = Flowable.fromPublisher(cacheSet.localPublisher(segment))
+               .doOnNext(ce -> seenKeys.add(ce.getKey()));
+         Flowable<CacheEntry<K, V>> loaderFlowable = Flowable.fromPublisher(persistenceManager.<K, V>publishEntries(IntSets.immutableSet(segment), k -> !seenKeys.contains(k), true, true, BOTH))
+               .map(me -> PersistenceUtil.convert(me, iceFactory));
+         return Flowable.concat(inMemoryPublisher, loaderFlowable);
+      }
+
+      @Override
+      public Publisher<CacheEntry<K, V>> localPublisher(IntSet segments) {
+         Set<K> seenKeys = new HashSet<>(cache.getAdvancedCache().getDataContainer().sizeIncludingExpired());
+         Flowable<CacheEntry<K, V>> inMemoryPublisher = Flowable.fromPublisher(cacheSet.localPublisher(segments))
+               .doOnNext(ce -> seenKeys.add(ce.getKey()));
+         Flowable<CacheEntry<K, V>> loaderFlowable = Flowable.fromPublisher(persistenceManager.<K, V>publishEntries(segments, k -> !seenKeys.contains(k), true, true, BOTH))
+               .map(me -> PersistenceUtil.convert(me, iceFactory));
+         return Flowable.concat(inMemoryPublisher, loaderFlowable);
+      }
```

Comment:
Okay, don't look at the PersistenceEntryStreamSupplier then ;)

I can change it to have the extra methods though.

Reply:
Actually, we don't want to do this. - We want to be invoking `cacheSet.localPublisher(int)` for the single segment one when possible - otherwise we are creating an immutable `IntSet` for every segment when parallelizing and an iterator. This isn't a big deal for the store, as such we only do the `IntSet` argument for it and the store performance will make this completely unnoticed. However when in memory we don't want to be allocating if we don't need to.

Reply:
I was more referring to `CacheSet.localPublisher(IntSet)`, but I could do what I am doing in the `ComposedSegmentedLoadWriteStore`

Reply:
I like your code here though @danberindei